File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/websocket/OdooWebSocketServer.kt
package com.fccmiddleware.edge.websocket

import com.fccmiddleware.edge.logging.AppLogger
import com.fccmiddleware.edge.adapter.common.IFccAdapter
import com.fccmiddleware.edge.buffer.dao.TransactionBufferDao
import com.fccmiddleware.edge.config.WebSocketDto
import io.ktor.server.application.install
import io.ktor.server.cio.CIO
import io.ktor.server.cio.CIOApplicationEngine
import io.ktor.server.engine.EmbeddedServer
import io.ktor.server.engine.embeddedServer
import io.ktor.server.plugins.origin
import io.ktor.server.routing.routing
import io.ktor.server.websocket.WebSockets
import io.ktor.server.websocket.pingPeriod
import io.ktor.server.websocket.timeout
import io.ktor.server.websocket.webSocket
import io.ktor.websocket.CloseReason
import io.ktor.websocket.Frame
import io.ktor.websocket.WebSocketSession
import io.ktor.websocket.close
import io.ktor.websocket.readText
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.time.Duration.Companion.seconds
import java.util.concurrent.ConcurrentHashMap

/**
 * Odoo backward-compatible WebSocket server.
 *
 * Mimics the DOMSRealImplementation Fleck-based WSS server protocol so Odoo POS
 * requires zero code changes. Listens on configurable port (default 8443).
 *
 * Per-connection pump status timer fires every 3 seconds, sending each
 * [FuelPumpStatusWsDto] individually to the specific client.
 *
 * All `mode` commands dispatched to [OdooWsMessageHandler].
 *
 * Security (S-005):
 *  - If [WebSocketDto.sharedSecret] is set, every connection must supply it in
 *    the `X-Api-Key` header or the upgrade is rejected with 4008 VIOLATED_POLICY.
 *  - Inbound command messages are rate-limited per [WebSocketDto.commandRateLimitPerMinute].
 *  - Every inbound command is logged with the client's remote IP for audit.
 */
class OdooWebSocketServer(
    private val transactionDao: TransactionBufferDao,
    private val serviceScope: CoroutineScope,
) {
    companion object {
        private const val TAG = "OdooWebSocketServer"

        /** S-005: header name for the shared-secret authentication token. */
        private const val API_KEY_HEADER = "X-Api-Key"
    }

    private val wsJson = Json {
        encodeDefaults = true
        ignoreUnknownKeys = true
        isLenient = true
    }

    @Volatile
    var config: WebSocketDto = WebSocketDto()
        private set

    /** Late-bound: wired when FCC adapter is available. */
    @Volatile
    internal var fccAdapter: IFccAdapter? = null

    /** Late-bound: wired when site config is available. */
    @Volatile
    internal var siteCode: String = ""

    private var server: EmbeddedServer<CIOApplicationEngine, CIOApplicationEngine.Configuration>? = null

    /**
     * Per-session tracking: remote host and rate-limit counters.
     * Rate-limit fields are mutated only from that session's single reader coroutine,
     * so no additional synchronisation is needed beyond the ConcurrentHashMap itself.
     */
    private data class ClientEntry(
        val remoteHost: String,
        var commandCountInWindow: Int = 0,
        var windowStartMs: Long = System.currentTimeMillis(),
    )

    /** All connected sessions mapped to their [ClientEntry]. */
    private val clients = ConcurrentHashMap<WebSocketSession, ClientEntry>()

    /** AP-016: Single shared pump status broadcast job instead of per-client coroutines. */
    private var sharedBroadcastJob: Job? = null

    private val messageHandler = OdooWsMessageHandler(
        transactionDao = transactionDao,
        wsJson = wsJson,
        broadcastToAll = ::broadcastToAll,
        getFccAdapter = { fccAdapter },
        getSiteCode = { siteCode },
    )

    fun wireFccAdapter(adapter: IFccAdapter?) {
        fccAdapter = adapter
    }

    fun wireSiteCode(code: String) {
        siteCode = code
    }

    fun reconfigure(newConfig: WebSocketDto) {
        val shouldRestart = server != null && config != newConfig
        config = newConfig
        if (shouldRestart) {
            start()
        }
    }

    fun start() {
        if (!config.enabled) {
            AppLogger.d(TAG, "WebSocket server disabled in config")
            return
        }
        server?.stop(1_000, 2_000)

        server = embeddedServer(CIO, port = config.port, host = config.bindAddress) {
            install(WebSockets) {
                pingPeriod = 15.seconds
                timeout = 30.seconds
                maxFrameSize = Long.MAX_VALUE
                masking = false
            }
            routing {
                webSocket("/") {
                    // S-005(a): extract client IP for auth and audit logging
                    val remoteHost = call.request.origin.remoteHost

                    // S-005(a): reject connections without the shared secret when configured
                    val secret = config.sharedSecret
                    if (!secret.isNullOrBlank()) {
                        val provided = call.request.headers[API_KEY_HEADER]
                        if (provided != secret) {
                            AppLogger.w(TAG, "Rejected unauthenticated WebSocket connection from $remoteHost")
                            close(CloseReason(CloseReason.Codes.VIOLATED_POLICY, "Unauthorized"))
                            return@webSocket
                        }
                    }

                    onClientConnected(this, remoteHost)
                }
            }
        }.also {
            it.start(wait = false)
            AppLogger.i(TAG, "WebSocket server started on ${config.bindAddress}:${config.port}")
        }

        // AP-016: Start single shared pump status broadcast coroutine
        sharedBroadcastJob?.cancel()
        sharedBroadcastJob = serviceScope.launch {
            sharedPumpStatusBroadcastLoop()
        }
    }

    fun stop() {
        // AP-016: Cancel shared pump status broadcast
        sharedBroadcastJob?.cancel()
        sharedBroadcastJob = null
        clients.clear()

        server?.stop(1_000, 2_000)
        server = null
        AppLogger.i(TAG, "WebSocket server stopped")
    }

    // -------------------------------------------------------------------------
    // Connection lifecycle
    // -------------------------------------------------------------------------

    private suspend fun onClientConnected(session: WebSocketSession, remoteHost: String) {
        if (clients.size >= config.maxConnections) {
            AppLogger.w(TAG, "Max connections (${config.maxConnections}) reached — rejecting $remoteHost")
            session.close(CloseReason(CloseReason.Codes.TRY_AGAIN_LATER, "Max connections reached"))
            return
        }

        clients[session] = ClientEntry(remoteHost = remoteHost)
        AppLogger.i(TAG, "Client connected from $remoteHost (total=${clients.size})")

        try {
            for (frame in session.incoming) {
                if (frame is Frame.Text) {
                    val text = frame.readText()
                    handleMessage(session, text, remoteHost)
                }
            }
        } catch (e: Exception) {
            AppLogger.d(TAG, "Client session ended ($remoteHost): ${e.message}")
        } finally {
            onClientDisconnected(session)
        }
    }

    private fun onClientDisconnected(session: WebSocketSession) {
        clients.remove(session)
        AppLogger.i(TAG, "Client disconnected (remaining=${clients.size})")
    }

    // -------------------------------------------------------------------------
    // Message routing
    // -------------------------------------------------------------------------

    private suspend fun handleMessage(session: WebSocketSession, text: String, remoteHost: String) {
        // S-005(b): rate limiting — sliding 60-second window per connection
        val limit = config.commandRateLimitPerMinute
        if (limit > 0) {
            val entry = clients[session]
            if (entry != null) {
                val now = System.currentTimeMillis()
                if (now - entry.windowStartMs > 60_000L) {
                    entry.commandCountInWindow = 0
                    entry.windowStartMs = now
                }
                entry.commandCountInWindow++
                if (entry.commandCountInWindow > limit) {
                    AppLogger.w(TAG, "Rate limit exceeded for $remoteHost — dropping message")
                    sendToSession(session, WsErrorResponse(message = "Rate limit exceeded"))
                    return
                }
            }
        }

        try {
            val json = wsJson.parseToJsonElement(text)
            if (json !is JsonObject) {
                sendToSession(session, WsErrorResponse(message = "Invalid message format"))
                return
            }

            val mode = json["mode"]?.jsonPrimitive?.content?.lowercase() ?: ""

            // S-005(c): audit-log every inbound command with client IP
            AppLogger.i(TAG, "WS cmd mode=$mode from=$remoteHost")

            when (mode) {
                "latest" -> messageHandler.handleLatest(session, json)
                "all" -> messageHandler.handleAll(session)
                "manager_update" -> messageHandler.handleManagerUpdate(session, json)
                "attendant_update" -> messageHandler.handleAttendantUpdate(session, json)
                "fuelpumpstatus" -> messageHandler.handleFuelPumpStatus(session)
                "fp_unblock" -> messageHandler.handleFpUnblock(session, json)
                "attendant_pump_count_update" -> messageHandler.handleAttendantPumpCountUpdate(session, json)
                "manager_manual_update" -> messageHandler.handleManagerManualUpdate(session, json)
                "add_transaction" -> messageHandler.handleAddTransaction(session, json)
                else -> {
                    AppLogger.w(TAG, "Unknown WS mode='$mode' from=$remoteHost")
                    sendToSession(session, WsErrorResponse(message = "Unknown mode '$mode'"))
                }
            }
        } catch (e: Exception) {
            AppLogger.e(TAG, "Error handling WebSocket message from $remoteHost", e)
            try {
                sendToSession(session, WsErrorResponse(message = "Internal server error"))
            } catch (_: Exception) { /* connection may be closing */ }
        }
    }

    // -------------------------------------------------------------------------
    // Pump status broadcast — single shared coroutine (AP-016)
    // -------------------------------------------------------------------------

    /**
     * AP-016: Single shared pump status broadcast loop.
     * Fetches pump statuses once per interval and fans out to all connected clients,
     * eliminating N duplicate FCC adapter queries when N clients are connected.
     */
    private suspend fun sharedPumpStatusBroadcastLoop() {
        val intervalMs = (config.pumpStatusBroadcastIntervalSeconds * 1000L).coerceAtLeast(1000L)
        while (serviceScope.isActive) {
            delay(intervalMs)
            if (clients.isEmpty()) continue
            try {
                val adapter = fccAdapter ?: continue
                val statuses = adapter.getPumpStatus()
                // Serialize each status DTO once, then reuse for all clients
                val frames = statuses.map { status ->
                    Frame.Text(wsJson.encodeToString(status.toWsDto()))
                }
                val deadSessions = mutableListOf<WebSocketSession>()
                for (session in clients.keys) {
                    try {
                        for (frame in frames) {
                            session.send(frame)
                        }
                    } catch (_: Exception) {
                        deadSessions.add(session)
                    }
                }
                deadSessions.forEach { onClientDisconnected(it) }
            } catch (e: Exception) {
                AppLogger.d(TAG, "Shared pump status broadcast failed: ${e.message}")
            }
        }
    }

    // -------------------------------------------------------------------------
    // Broadcast helpers
    // -------------------------------------------------------------------------

    /**
     * Send a typed message to ALL connected clients.
     * Used by handlers for transaction_update broadcasts.
     */
    internal suspend fun broadcastToAll(type: String, data: Any?) {
        val payload = wsJson.encodeToString(
            kotlinx.serialization.json.buildJsonObject {
                put("type", kotlinx.serialization.json.JsonPrimitive(type))
                if (data != null) {
                    // data is already a JsonElement from the handler
                    if (data is kotlinx.serialization.json.JsonElement) {
                        put("data", data)
                    }
                }
            }
        )
        val deadSessions = mutableListOf<WebSocketSession>()
        for (session in clients.keys) {
            try {
                session.send(Frame.Text(payload))
            } catch (_: Exception) {
                deadSessions.add(session)
            }
        }
        deadSessions.forEach { onClientDisconnected(it) }
    }

    private suspend inline fun <reified T> sendToSession(session: WebSocketSession, data: T) {
        try {
            val payload = wsJson.encodeToString(data)
            session.send(Frame.Text(payload))
        } catch (e: Exception) {
            AppLogger.d(TAG, "Failed to send to session: ${e.message}")
        }
    }
}

File: src/desktop-edge-agent/src/FccDesktopAgent.Core/WebSocket/OdooWebSocketServer.cs
using System.Collections.Concurrent;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using FccDesktopAgent.Core.Adapter.Common;
using FccDesktopAgent.Core.Buffer;
using FccDesktopAgent.Core.Security;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace FccDesktopAgent.Core.WebSocket;

/// <summary>
/// Odoo backward-compatible WebSocket server using ASP.NET Core Kestrel.
///
/// Mimics the DOMSRealImplementation Fleck-based WSS server protocol so Odoo POS
/// requires zero code changes. Listens on configurable port (default 8443).
///
/// Per-connection pump status timer fires every 3 seconds, sending each
/// <see cref="FuelPumpStatusWsDto"/> individually to the specific client.
/// </summary>
public sealed class OdooWebSocketServer : IDisposable
{
    private readonly IServiceScopeFactory _scopeFactory;
    private readonly ILogger<OdooWebSocketServer> _logger;

    private readonly ConcurrentDictionary<System.Net.WebSockets.WebSocket, CancellationTokenSource> _clients = new();

    private readonly JsonSerializerOptions _jsonOptions = new()
    {
        PropertyNamingPolicy = null, // exact field names from [JsonPropertyName]
        WriteIndented = false,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
    };

    // P-DSK-012/020: Single shared pump status broadcast loop instead of per-connection loops.
    private CancellationTokenSource? _broadcastLoopCts;
    private Task? _broadcastLoopTask;
    private readonly object _broadcastLoopLock = new();

    /// <summary>Late-bound: wired when FCC adapter becomes available.</summary>
    public IFccAdapter? FccAdapter { get; set; }

    /// <summary>Late-bound: wired when pump status service becomes available.</summary>
    public IPumpStatusService? PumpStatusService { get; set; }

    /// <summary>WebSocket server configuration.</summary>
    public WebSocketServerOptions Options { get; set; } = new();

    public OdooWebSocketServer(
        IServiceScopeFactory scopeFactory,
        ILogger<OdooWebSocketServer> logger)
    {
        _scopeFactory = scopeFactory;
        _logger = logger;
    }

    /// <summary>
    /// Handles a single WebSocket connection. Called by Kestrel middleware.
    /// </summary>
    public async Task HandleConnectionAsync(
        System.Net.WebSockets.WebSocket webSocket,
        CancellationToken ct)
    {
        if (_clients.Count >= Options.MaxConnections)
        {
            _logger.LogWarning("Max WebSocket connections ({Max}) reached — rejecting client", Options.MaxConnections);
            await webSocket.CloseAsync(
                WebSocketCloseStatus.PolicyViolation,
                "Max connections reached",
                ct);
            return;
        }

        var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        _clients[webSocket] = cts;
        _logger.LogInformation("WebSocket client connected (total={Total})", _clients.Count);

        // P-DSK-012/020: Start the shared pump status broadcast loop when the first client connects
        EnsureSharedBroadcastLoopStarted(ct);

        try
        {
            await ReceiveLoopAsync(webSocket, cts.Token);
        }
        catch (OperationCanceledException) { }
        catch (WebSocketException) { }
        finally
        {
            cts.Cancel();
            _clients.TryRemove(webSocket, out _);
            _logger.LogInformation("WebSocket client disconnected (remaining={Remaining})", _clients.Count);

            if (webSocket.State is WebSocketState.Open or WebSocketState.CloseReceived)
            {
                try
                {
                    await webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Goodbye", CancellationToken.None);
                }
                catch { /* closing — safe to ignore */ }
            }
        }
    }

    // ── Receive loop ────────────────────────────────────────────────────────

    /// <summary>
    /// S-DSK-020: Maximum allowed WebSocket message size (1 MB).
    /// Messages exceeding this limit are rejected and the connection is closed.
    /// </summary>
    private const int MaxMessageSize = 1 * 1024 * 1024;

    private async Task ReceiveLoopAsync(System.Net.WebSockets.WebSocket ws, CancellationToken ct)
    {
        var buffer = new byte[4096];
        while (!ct.IsCancellationRequested && ws.State == WebSocketState.Open)
        {
            var sb = new StringBuilder();
            WebSocketReceiveResult result;

            do
            {
                result = await ws.ReceiveAsync(buffer, ct);
                if (result.MessageType == WebSocketMessageType.Close)
                    return;
                sb.Append(Encoding.UTF8.GetString(buffer, 0, result.Count));

                // S-DSK-020: Guard against unbounded memory allocation
                if (sb.Length > MaxMessageSize)
                {
                    _logger.LogWarning(
                        "WebSocket message exceeded maximum size ({MaxSize} bytes) — closing connection",
                        MaxMessageSize);
                    await ws.CloseAsync(
                        WebSocketCloseStatus.PolicyViolation,
                        "Message too large",
                        ct);
                    return;
                }
            }
            while (!result.EndOfMessage);

            var message = sb.ToString();
            await HandleMessageAsync(ws, message, ct);
        }
    }

    // ── Message routing (switch on `mode`) ──────────────────────────────────

    private async Task HandleMessageAsync(
        System.Net.WebSockets.WebSocket ws,
        string message,
        CancellationToken ct)
    {
        try
        {
            using var doc = JsonDocument.Parse(message);
            var root = doc.RootElement;
            var mode = root.TryGetProperty("mode", out var m)
                ? m.GetString()?.ToLowerInvariant() ?? ""
                : "";

            var handler = new OdooWsMessageHandler(
                _scopeFactory, _logger, _jsonOptions, BroadcastToAllAsync);

            switch (mode)
            {
                case "latest":
                    await handler.HandleLatestAsync(ws, root, ct);
                    break;
                case "all":
                    await handler.HandleAllAsync(ws, ct);
                    break;
                case "manager_update":
                    await handler.HandleManagerUpdateAsync(ws, root, ct);
                    break;
                case "attendant_update":
                    await handler.HandleAttendantUpdateAsync(ws, root, ct);
                    break;
                case "fuelpumpstatus":
                    await handler.HandleFuelPumpStatusAsync(ws, PumpStatusService, ct);
                    break;
                case "fp_unblock":
                    await handler.HandleFpUnblockAsync(ws, root, FccAdapter, ct);
                    break;
                case "attendant_pump_count_update":
                    await handler.HandleAttendantPumpCountUpdateAsync(ws, root, ct);
                    break;
                case "manager_manual_update":
                    await handler.HandleManagerManualUpdateAsync(ws, root, ct);
                    break;
                case "add_transaction":
                    break; // No-op: FCC is source of truth
                default:
                    await SendAsync(ws, new { status = "error", message = $"Unknown mode '{mode}'" }, ct);
                    break;
            }
        }
        catch (JsonException)
        {
            await SendAsync(ws, new { status = "error", message = "Invalid JSON" }, ct);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error handling WebSocket message");
            try { await SendAsync(ws, new { status = "error", message = "Internal server error" }, ct); }
            catch { /* connection closing */ }
        }
    }

    // ── Pump status broadcast (shared site-wide loop) ───────────────────────

    // P-DSK-012/020: Single shared broadcast loop that queries pump status once
    // and sends to all connected clients, instead of per-connection loops.
    private void EnsureSharedBroadcastLoopStarted(CancellationToken ct)
    {
        lock (_broadcastLoopLock)
        {
            if (_broadcastLoopTask is not null && !_broadcastLoopTask.IsCompleted)
                return;

            _broadcastLoopCts?.Dispose();
            _broadcastLoopCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            _broadcastLoopTask = SharedPumpStatusBroadcastLoopAsync(_broadcastLoopCts.Token);
        }
    }

    private async Task SharedPumpStatusBroadcastLoopAsync(CancellationToken ct)
    {
        var interval = TimeSpan.FromSeconds(Math.Max(1, Options.PumpStatusBroadcastIntervalSeconds));
        await Task.Delay(interval, ct); // Initial delay before first broadcast

        while (!ct.IsCancellationRequested)
        {
            if (_clients.IsEmpty)
            {
                await Task.Delay(interval, ct);
                continue;
            }

            try
            {
                var svc = PumpStatusService;
                if (svc is not null)
                {
                    var result = await svc.GetPumpStatusAsync(null, ct);
                    foreach (var status in result.Pumps)
                    {
                        var dto = status.ToWsDto();
                        await BroadcastToAllAsync("FuelPumpStatus", dto, ct);
                    }
                }
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                _logger.LogDebug(ex, "Shared pump status broadcast failed");
            }

            await Task.Delay(interval, ct);
        }
    }

    // ── Broadcast to all ────────────────────────────────────────────────────

    // P-DSK-025: Send to all clients in parallel with bounded concurrency so one slow
    // socket doesn't delay delivery to every other terminal.
    internal async Task BroadcastToAllAsync(string type, object? data, CancellationToken ct)
    {
        var payload = JsonSerializer.Serialize(new { type, data }, _jsonOptions);
        var bytes = Encoding.UTF8.GetBytes(payload);
        var dead = new ConcurrentBag<System.Net.WebSockets.WebSocket>();
        var snapshot = _clients.Keys.ToArray();

        await Parallel.ForEachAsync(snapshot, new ParallelOptions
        {
            MaxDegreeOfParallelism = Math.Min(snapshot.Length, 10),
            CancellationToken = ct,
        }, async (ws, token) =>
        {
            if (ws.State != WebSocketState.Open) { dead.Add(ws); return; }
            try
            {
                await ws.SendAsync(bytes, WebSocketMessageType.Text, true, token);
            }
            catch
            {
                dead.Add(ws);
            }
        });

        foreach (var ws in dead)
        {
            if (_clients.TryRemove(ws, out var cts))
                cts.Cancel();
        }
    }

    private async Task SendAsync(System.Net.WebSockets.WebSocket ws, object data, CancellationToken ct)
    {
        if (ws.State != WebSocketState.Open) return;
        var json = JsonSerializer.Serialize(data, _jsonOptions);
        var bytes = Encoding.UTF8.GetBytes(json);
        await ws.SendAsync(bytes, WebSocketMessageType.Text, true, ct);
    }

    public void Dispose()
    {
        _broadcastLoopCts?.Cancel();
        _broadcastLoopCts?.Dispose();

        foreach (var (ws, cts) in _clients)
        {
            cts.Cancel();
            cts.Dispose();
        }
        _clients.Clear();
    }
}

/// <summary>
/// Configuration for the Odoo WebSocket server.
/// </summary>
public sealed class WebSocketServerOptions
{
    public const string SectionName = "WebSocket";

    public bool Enabled { get; set; }
    public int Port { get; set; } = 8443;
    public int MaxConnections { get; set; } = 10;
    public int PumpStatusBroadcastIntervalSeconds { get; set; } = 3;

    /// <summary>Whether to enable TLS (WSS) on the WebSocket port. Requires <see cref="CertificatePath"/>.</summary>
    public bool UseTls { get; set; }

    /// <summary>Absolute path to a PFX/PKCS#12 certificate file for TLS.</summary>
    public string? CertificatePath { get; set; }

    /// <summary>Password for the PFX certificate file.</summary>
    [SensitiveData]
    public string? CertificatePassword { get; set; }
}
